docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,341 @@
+evolution.
+2. The separation between the client and server interfaces will allow us to insert permission checking code in between
+the two which checks if one module is authorized to send the specified `Msg` to the other module providing a proper
+object capability system.
+
+This mechanism has the added benefits of:
+- reducing boilerplate through code generation, and
+- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC 
+
+### Inter-module Communication
+
+In order for code to use the `Client` interfaces generated by the protobuf compiler, a `grpc.ClientConn` implementation
+is needed. We introduce a new type, `ModuleKey`, to serve this role which we can conceptualize as the "private key"
+corresponding to a module account.
+
+Whereas external clients use their private key to sign transactions containing `Msg`s where they are listed as signers,
+modules use their `ModuleKey` to send `Msg`s where they are listed as the sole signer to other modules. For example, modules
+could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another
+account.
+
+`QueryClient`s could also be made with `ModuleKey`s, except that authentication isn't required.
+
+Here's an example of a hypothetical module `foo` interacting with `x/bank`:
+```go
+package foo
+
+func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
+  bankQueryClient := bank.NewQueryClient(fooMsgServer.moduleKey)
+  balance, err := bankQueryClient.Balance(&bank.QueryBalanceRequest{Address: fooMsgServer.moduleKey.Address(), Denom: "foo"})
+  
+  ...
+
+  bankMsgClient := bank.NewMsgClient(fooMsgServer.moduleKey)
+  res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: fooMsgServer.moduleKey.Address(), ...})
+
+  ...
+}
+```
+
+### `ModuleKey`s and `ModuleID`s
+
+A `ModuleKey` can be thought of as a "private key" for a module account and a `ModuleID` can be thought of as the
+corresponding "public key". From [the ADR 028 draft](https://github.com/cosmos/cosmos-sdk/pull/7086), modules can have both a root module account and any number of sub-accounts
+or derived accounts that can be used for different pools (ex. staking pools) or managed accounts (ex. group
+accounts). We can also think of module sub-accounts as similar to derived keys - there is a root key and then some
+derivation path. `ModuleID` is a simple struct which contains the module name and optional "derivation" path,
+and forms its address based on the `AddressHash` method from [the ADR 028 draft](https://github.com/cosmos/cosmos-sdk/pull/7086):
+
+```go
+type ModuleID struct {
+  ModuleName string
+  Path []byte
+}
+
+func (key ModuleID) Address() []byte {
+  return AddressHash(key.ModuleName, key.Path)
+}
+```
+
+In addition to being able to generate a `ModuleID` and address, a `ModuleKey` contains a special function closure called
```

Comment:
see `RootModuleKey` and `DerivedModuleKey` examples below, the `ModuleId` can be generated from the `moduleName` (and `path` in case of `DerivedModuleKey`), but I agree this would be great to make this clearer.

Reply:
Also, as outlined in this ADR, we can think of the `ModuleKey` as a private key and `ModuleId` as a public key.
So similar to our `PrivKey` interface which defines `PubKey()` method, `ModuleKey` defines a `ModuleId()` method.